doc/designs/event-queue.md

```diff
@@ -0,0 +1,181 @@
+    that need customized destruction when the event structure itself is to
+    be destroyed.
+
+The event itself is designed for a single synchronous thread, i.e. cannot be
+shared by multiple threads.  The diverse objects it refers to may, however,
+be shared by multiple threads, at the discretion of the functions in the
+method structure.
+
+Once populated, the event type, the references to event context, and the
+reference to the destructor function are considered immutable, up until the
+event structure is destroyed.
+
+The reference to the auxilliary identifying material or to the payload,
+however, are considered mutable.  Any event handler may "steal" them and
+replace the reference to them in the event structure with NULL.  Stealing
+must be done with much care.
+
+Events may be dynamically allocated.  In that case, ownership is passed to
+the event dispatch functionality (see below), and the event will be
+automatically destroyed when not needed any more.
+
+Events may also be embedded in another structure, as a static variable, or
+as an local variable in a function.  In that case, owndership remains with
+that structure or that variable.
+
+### Event queue
+
+An event queue is a structure holding the following:
+
+-   A reference to a method structure to handle the event queue.
+-   A reference to auxilliary data for the methods.
+-   An array of subscribers, which are simple callbacks taking a reference
+    to an event structure, along with auxilliary arguments.
+-   Buckets of events, one bucket for each priority level.  The number of
+    buckets depends on how many priority levels are implemented by the
+    functions of the method structure.
+    If there are zero priority levels, there are no buckets at all, and
+    events "added" to such a queue structure are always dispatched
+    immediately, if the function used to "add" the event supports that.
+
+For each event added to the queue, there is associated data:
+
+-   *[optional]* A time value, to indicate when the event is to be fired
+    off (i.e. when it's going to be passed to subscribers)
+
+    This may be zero to indicate that it should be fired off as soon as
+    possible, and will otherwise be compared to the current time to figure
+    out if it should be fired off or should remain in the event queue for a
+    bit longer.
+
+    The time value allows for time delay events, where the subscriber that
+    listens to them can simply generate a new time delay event, thereby
+    ensuring that they happen at regular intervals without any extra help
+    from outside calls.
+
+#### Associated structures and services
+
+The ctrl() function in the method structure of an event queue may manipulate
+the auxilliary data, as well as perform extraordinary actions that affect
+other structures, such as other event queues, as well as the event queue
+that such calls originate from.
+
+Similarly, subscribers may affect other structures, even other event queues,
```

Comment:
How many event queue are we expecting to need? I was expecting one per SSL_CTX...or SSL_EVENT_CTX...or wherever we put it.